### ISSUE TITLE
Displaying generic image collections

### ISSUE THREAD

**Issue by d-v-b** [napari/napari#661]
## 🚀 Feature
Napari currently assumes that within a viewer, all images occupy the same physical space. This assumption is invalid for a lot of collections of images, e.g. sample images in `skimage.data`:

```python
import skimage.data as skid
list_of_images = [skid.astronaut(), skid.clock(), skid.brick(), skid.binary_blobs(length=50,n_dim=3)]
# This would be intuitive, but it doesn't work -- array dimensions don't align
# napari.view_image(list_of_images)

# This works, but the logic of the viewer is weird -- the images are layered, seeing them individually is hard
v = napari.viewer.Viewer()
[v.add_image(d) for d in list_of_images]
```

This renders all the images from `list_of_images` in the same coordinate space, and gives them all a Z axis slider even though most images are 2D. Seeing these images individually thus requires extra effort (layers above and below must be masked), and there is are vestigial GUI element (the z-slider, and the "roll displayed axes" button).

Ideally (from my perspective) calling `napari.view_image(list_of_images)` would generate a single image layer with a slider controlling which image is rendered, and additional sliders appearing as needed per image (e.g., the blobs image is 3d, so it needs an extra dimensions slider). 

## Motivation

For my work, I need to view a large collection of 3D/4D images that have different dimensions and are from totally different spaces, so it doesn't make sense for them to be overlayed, and it doesn't make sense for them to share dimension sliders. But I would like to quickly browse them!  

These images are human-annotated training data for a machine learning application; I suspect that lots of training data for 2D/3D ML will be collections of arrays with different sizes from different spaces, so the issue I'm facing might be fairly general.

## Pitch

A generalized image viewer for collections of independent images. This could be accomplished via an internal flag that determines whether the viewer should render images in the same space or not; this flag could be set based on whether a) the user supplies a kwarg, e.g. `napari.view_image(data_3d, shared_coordinates=False)` b) whether the user supplies a collection of images with different sizes /shapes, c) some other heuristic.

## Additional context

This problem reminds me of the perennial "is this last axis RGB, or is it just an axis that happens to have length== 3" problem. I think the solution there is to assume the general case, and allow the user to supply flags for the special case. The same approach might work here, once we figure out which case is specific and which is general. From an abstract perspective, the `NDArray` is a special case of `collection of arrays`, but empirically I'm not sure what which type of data people are viewing more often.

**Comment by sofroniewn** [comment]
This also relates to #605. I need to think about it more, but it would be pretty straight forward now if you can package everything inside one large dask array, as dask image imread does. If the biggest dim image have dim 5 you'd get 3 sliders, but it should still be fine.

Can you try packaging `list_of_images` inside a dask array?

**Comment by d-v-b** [comment]
putting `list_of_images` in a dask array requires first putting it in a numpy array, and this still generates errors when passed to `napari.view_image`. The first error I got was in calculating the contrast limits, and the second error was in interpolation during construction of the thumbnail image.

**Comment by jni** [comment]
@d-v-b I think #605 captures this quite well. Having said that, using layers, then switching to grid mode is probably pretty effective as a "for-now" workaround?

I don't know what you think is the correct behaviour for mixed 2D and 3D images, though. I think it would look pretty terrible to have one of the sliders popping in and out of existence...?

**Comment by d-v-b** [comment]
> @d-v-b I think #605 captures this quite well. Having said that, using layers, then switching to grid mode is probably pretty effective as a "for-now" workaround?

Yeah, at the moment I can work around the issue. In the worst case I could define a dask array that does padding to the maximal dimensions, thereby ensuring that all axes are identical.

> I don't know what you think is the correct behaviour for mixed 2D and 3D images, though. I think it would look pretty terrible to have one of the sliders popping in and out of existence...?

I agree that "surprise sliders" should be avoided, but it's also bad to have a slider that doesn't do anything for the current image. What if sliders were enabled/disabled based on the dimensionality of the current active layer? The number of sliders would still be determined by the maximal dimensionality across layers and (thus constant while viewing a fixed set of layers), but sliders would get greyed out if you select a 2D image layer. Might be worth trying?

**Comment by sofroniewn** [comment]
The dask array padding works quite nicely when they all have the same dims - and this is certainly something we could consider doing under the hood if it was desirable - people could pass a list of images and we'd take care of the rest, for example:

```python
from skimage import data
import dask.array as da
import numpy as np
import napari


list_of_images = [data.moon(), data.clock(), data.coins()]
max_shape = np.max([im.shape for im in list_of_images], axis=0)
full_array = da.concatenate([da.pad(da.from_array([im]),
                                    [(0, 0)] + [(i//2, (i+1)//2)for i in np.subtract(max_shape, im.shape)],
                                     mode='constant') for im in list_of_images])
napari.view_image(full_array)
```

We could even do this for cases where the dims weren't the same if really wanted. I'd probably put this intercept in the `viewer.add_image` function so that we didn't change the `Image` model, but just did this conversion there. It would also mean we could move towards supporting #605 more easily.

**Comment by d-v-b** [comment]
I'm not sure I like the idea of napari padding all images in a layer to make them the same size... I would not expect that an image viewer would add data to my images. E.g., padding leads to napari erroneously reporting that the outside of the image has value 0. And when we start to think about arrays with differing dimensionality, the memory cost of padding a 2D array up to 5D might be prohibitive. 

More generally, I think it would be better to avoid coercing input data with diverse size / shape into an NDarray. The axes in the skimage sample images are not the same, so putting these images in an NDarray via padding just seems weird. To me, the natural data structure for these sample images is just a 1D collection, e.g. a list. If the `image_layer` is committed to only supporting NDarrays, maybe this use case is just out-of-scope for that layer?

**Comment by jni** [comment]
> If the image_layer is committed to only supporting NDarrays, maybe this use case is just out-of-scope for that layer?

I wouldn't say *committed* to *only*, but that is the primary use case, and I am reluctant to overly complicate the code to support secondary use cases. If we can come up with a simple, transparent solution that works, I'm all for it. Part of the reason that it really complicates things is that now if you switch to 3D, you *must* instantiate a "cube" of data. Or if you roll the axes — what happens then? Maybe we need a different kind of interaction here.

**Comment by d-v-b** [comment]
This problem surfaced again for me -- I wanted to view the output of a bunch of different filters applied to the same data (pursuant to playing with [this example](https://github.com/guiwitz/napari_demo/blob/master/napari_ilastik.ipynb) of the ilastik-style functionality within napari). I.e.: 
```python
im_gauss = skimage.filters.gaussian(image,sigma = 10, preserve_range = True)
im_gauss2 = skimage.filters.gaussian(image,sigma = 20, preserve_range = True)
im_frangi = skimage.filters.frangi(image)
im_prewitt = skimage.filters.prewitt(image)
im_meijering = skimage.filters.meijering(image)
im_gauss_laplace = skimage.filters.laplace(skimage.filters.gaussian(image,sigma = 5,preserve_range=True),ksize=10)
im_gradient = skimage.filters.rank.gradient(image,skimage.morphology.disk(8))
im_entropy = skimage.filters.rank.entropy(image,skimage.morphology.disk(8))
im_roberts = skimage.filters.roberts(skimage.filters.gaussian(image,sigma = 5,preserve_range=True))

all_filt = [im_gauss,im_gauss2,im_frangi,im_prewitt,im_meijering,im_gauss_laplace,im_gradient,im_entropy,im_roberts]
names = ['Gauss $\sigma$=10', 'Gauss $\sigma$=20','Frangi','Prewitt','Meijering','Gauss+Laplace', 'Gradient',
        'Entropy','Roberts']
```
Since all the images in `all_filt` are the same size, they can be put in a numpy array, which would in principle be nice for viewing with napari. However napari assigns a single LUT to the entire array, so no filtered image is visible -- the dynamic range of the gaussian-filtered images is totally different from the dynamic range of the edge-filtered images. 

The alternative is to make a single layer per image, but then to view a single image, you have to hide all the images on top, which can amount to `~O(num_images)` clicking. 

I think this use case, and the one I have detailed above, would not be too hard to accommodate: add an `add_image_collection()` method to the `Viewer` that takes a 1D collection of images. Adding this layer to the viewer creates a slider that indexes the the collection. If the collection contains 3D or higher images, additional sliders are created, but these sliders are disabled when they do not match the dimensionality of the currently displayed image. Likewise for the rollaxes and 3D viewer buttons: these would only be active when the currently displayed image from the collection supports them.

This `image_collection` layer would move some of the logic of multiple image layers to a single layer. This would be fine on its own, but I don't think it would work with the current napari model where all the layers are in the same (implictly defined) viewer space. So I think sorting out https://github.com/napari/napari/issues/659 is a prerequisite for sorting out this issue...

**Comment by jni** [comment]
@d-v-b indeed, I do think need to figure out our real-space interaction before tackling image collections.

I wonder if this specific use case, though, would be solved by a more powerful layer interaction model. These images all do refer to the same space, after all, so I think layers is the right conceptual model. People working with multispectral data have noticed that napari begins to fray at the edges with too many layers, so it’s something that needs improvement in various scenarios.
